MuvrKit/MKSensorDataEncoder.swift

```diff
@@ -0,0 +1,191 @@
+    private var offset: UInt64 = 0
+    
+    public init(data: NSMutableData) {
+        self.data = data
+    }
+    
+    public func writeData(data: NSData, offset: UInt64?) {
+        if let offset = offset {
+            self.data.replaceBytesInRange(NSRange(location: Int(offset), length: data.length), withBytes: data.bytes, length: data.length)
+        } else {
+            self.data.appendData(data)
+        }
+    }
+    
+    public func close() {
+        
+    }
+    
+    
+}
+
+///
+/// Provides direct-to-file encoding for the ``MKSensorData``
+///
+public final class MKSensorDataEncoder {
+    /// the target
+    private let target: MKSensorDataEncoderTarget
+    /// The types of data contained in the column vectors (or spans of column vectors)
+    private let types: [MKSensorDataType]
+    /// The sample dimension
+    private let dimension: Int
+    /// The samples per second
+    private let samplesPerSecond: UInt
+    /// the sample count
+    private var sampleCount: UInt32
+    /// the first sample date
+    private(set) public var startDate: NSDate?
+    /// the last written sample
+    private var lastSample: [Float]?
+    ///
+
+    ///
+    /// Initializes this encoder, setting the target, types, start and sampling rate
+    ///
+    public init(target: MKSensorDataEncoderTarget, types: [MKSensorDataType], samplesPerSecond: UInt) {
+        self.types = types
+        self.samplesPerSecond = samplesPerSecond
+        self.target = target
+        self.sampleCount = 0
+        self.dimension = types.reduce(0) { sum, type in return sum + type.dimension }
+
+        let emptyHeader = [UInt8](count: 16 + 4 * types.count, repeatedValue: 0)
+        target.writeData(NSData(bytes: emptyHeader, length: emptyHeader.count), offset: nil)
+    }
+    
+    ///
+    /// Append one "row" containing the sampled values
+    /// - parameter sample: the sample
+    ///
+    public func append(sample: [Float], sampleDate: NSDate) {
+        
+        func appendSample(sample: [Float]) {
+            sampleCount += UInt32(sample.count / dimension)
+            target.writeData(NSData(bytes: sample, length: sizeof(Float) * sample.count), offset: nil)
+        }
+        
+        /// generate missing samples between ``lastSample`` and new ``sample``
+        /// using linear interpolation
+        func generateSamples(count: Int) -> [Float] {
+            var samples = [Float](count: count * dimension, repeatedValue: 0)
+            for i in 0..<dimension {
+                let first = sample[i]
+                let last  = lastSample?[i] ?? first
+                let ds = Float(first - last) / Float(count + 1)
+                for j in 0..<count {
+                    samples[i + dimension * j] = last + ds * Float(j + 1)
+                }
+            }
+            samples.appendContentsOf(sample)
+            return samples
+        }
+
+        if sample.count == 0 {
+            fatalError("Empty samples.")
```

Comment:
It should not happen because the ``append`` method is not called when there is no data to write.